File: src/tsv_output.rs
use std::io;

use crate::*;

/// Supports writing tab-separated value output to any writer.
/// Typically the output will go to stdout or to a file.
pub struct TabSeparatedValuesOutput<D: io::Write> {
    dest: D,
    columns: Vec<Column>,
}

impl<D: io::Write> TabSeparatedValuesOutput<D> {
    pub fn new(dest: D) -> Self {
        TabSeparatedValuesOutput {
            dest,
            columns: vec![],
        }
    }
}

impl<D: std::io::Write> Output for TabSeparatedValuesOutput<D> {
    fn start(&mut self, columns: &[Column]) -> Result<()> {
        self.columns.extend_from_slice(columns);

        write_header_row(&mut self.dest, columns)
    }

    fn add_row(&mut self, date: &str, column_to_lines_map: &ColumnToLinesMap) -> Result<()> {
        write_row(&mut self.dest, date, &self.columns, column_to_lines_map)
    }

    fn finish(&mut self) -> Result<()> {
        eprintln!("\nCopy and paste the above output into your favourite spreadsheet software and make a graph.");

        Ok(())
    }
}

fn write_header_row(write: &mut dyn std::io::Write, columns: &[crate::Column]) -> Result<()> {
    // To get correct tab alignment, pad with spaces in place of a date
    // on the first row
    write!(write, "{}", " ".repeat("YYYY-MM-DD".len()))?;

    // Now write the columns
    for ext in columns {
        write!(write, "\t{}", ext)?;
    }

    // ... and finish with a newline
    writeln!(write)?;

    Ok(())
}

fn write_row(
    write: &mut dyn std::io::Write,
    date: &str,
    columns: &[crate::Column],
    column_to_lines_map: &crate::ColumnToLinesMap,
) -> Result<()> {
    // Date
    write!(write, "{}", date)?;

    // Line count information
    for column in columns {
        write!(write, "\t{}", column_to_lines_map.get(column).unwrap_or(&0))?;
    }

    // ... and finish with a newline
    writeln!(write)?;

    Ok(())
}

File: src/git_repo_language_trends/_internal/tsv_output.py
import sys

def write_header_row(writer, columns):
    # To get correct tab alignment, pad with spaces in place of a date
    # on the first row
    writer.write("          ")  # for YYYY-MM-DD

    # Now write the columns
    for column in columns:
        writer.write(f"\t{column}")

    # ... and finish with a newline
    writer.write("\n")


def write_row(writer, columns, date, column_to_lines_dict):
    # Date
    writer.write(date)

    # Line count information
    for column in columns:
        writer.write(f"\t{column_to_lines_dict.get(column, 0)}")

    # ... and finish with a newline
    writer.write("\n")


class TabSeparatedValuesStdoutOutput:
    """
    Implemented this to support formating the collected data in some kind of output format.
    Example output formats could be stdout printouts, a .tsv file, or a .png file with a graph.
    """

    # Called in the beginning of processing,
    # to announce what columns to use in the output.
    def start(self, columns):
        write_header_row(sys.stdout, columns)

    # Called each time a commit has been analyzed,
    # and it is time to present the data in the commit.
    def add_row(self, columns, date, column_to_lines_dict):
        write_row(sys.stdout, columns, date, column_to_lines_dict)

    # Called when the processing of commits is complete.
    # This is a good time to write output files to disk.
    def finish(self):
        pass
